### DISCUSSION TITLE
Best libraries to use for fitting

### DISCUSSION THREAD

**Discussion by eigenforesi** [community/community#132813] [Programming Help]
### Statistical Analysis/Fitting with non-template fit function

Hello,

I would like to do some statistical analysis in python but am not sure of the best tools to do this.  I would use something like pyhf and cabinetry, however my desired fit function doesn't seem to be one of the template functions as I have a somewhat special distribution.  I need to fit a distribution (using a maximum log likelihood fit) to the function y = a(1 + k(b*cos(2x) + c*sin(2x))).  My input is a ROOT TH1 histogram (or possibly multiple histograms) and I would like to extract the parameters, their uncertainties, the maximum log likelihood, and other statistical information about correlations and covariance and other info if possible.

I know I could do this using basic ROOT methods or scipy, however I need a library that allows me to do more sophisticated statistical analysis, for example I need to construct a model to perform a simultaneous fit to multiple histograms, forcing some of the parameters (b and c) to have the same values across histograms while a and k are allowed to be different.

I tried to do this using RooFit however it is being very weird with normalization and the a value I get is completely wrong and has a huge uncertainty.  I would like to do this fit/analysis without normalizing the fit function (in RooFit or another library, whichever works best).

If anyone could give suggestions on what libraries are my best options to do this, I would very much appreciate this!

Thank you!

Matt

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by ghost** [comment]
Hi, take a look at the pymc3 or numpy

**Reply by eigenforesi** [reply]
Does scipy do maximum likelihood or simultaneous fits of multiple histograms though?

**Reply by ghost** [reply]
I don’t know for sure about this, it’s better to look on the Internet, but it seems possible

**Comment by davevad93** [comment]
Hi @stormhusky22 ,

First of all I'm no Data Science expert, my focus is fullstack development, but I know some statistical analysis libraries thanks to some courses I've taken in the past. Considering what you are looking for, I think the following tools could be useful along with  `numpy`, `scipy`, `matplotlib` and `seaborn`:

1. [uproot](https://pypi.org/project/uproot/) for reading and writing `ROOT` files in `Python` and `numpy`.

2. [iminuit](https://pypi.org/project/iminuit/) to perform the fit and extract parameters, here are some features:

   - Builtin cost functions for statistical fits to N-dimensional data.
   - Template fits with error propagation.
   - Unbinned and binned maximum-likelihood + extended versions.
   - Least-squares.
   - Gaussian penalty terms for parameters.
   - Cost functions can be combined by adding them: total_cost = cost_1 + cost_2
   - Visualization of the fit in Jupyter notebooks.

Surely someone more knowledgeable than me in this field could give more precise advice on this subject, nonetheless I hope you find it helpful.
